{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "83e076fb-dd34-4592-b712-eca2f9a6541f",
   "metadata": {},
   "source": [
    "# Differential expression analysis\n",
    "of RNA-seq data received 2026-01-13\n",
    "\n",
    "## Results\n",
    "\n",
    "### PCA\n",
    "- MGG152_UNC1 a clear outlier\n",
    "- After removal, PC1 strongly correlated with sequencing depth (83%) variance\n",
    "- No obvious association with treatment in PCs 2 or 3.\n",
    "\n",
    "### Differential Expression\n",
    "- DE for IDHi:\n",
    "    |gene  |  baseMean  |  l2FC      |      lfcSE   |    p       |        q   |\n",
    "    |------|------------|------------|--------------|------------|------------|\n",
    "    |CBLN1 |  175.0304\t|-3.06166e-05|\t0.001442798\t|3.860136e-06|\t0.04981505|\n",
    "- DE for KMT5Ai:\n",
    "    |gene|baseMean|l2FC     | lfcSE   |   p       |      q   |\n",
    "    |----|--------|---------|---------|-----------|----------|\n",
    "    |H1-0|382.3084|0.7500887|0.1811774|1.43793e-06|0.01855649|\n",
    "- No significant hits for vorasidenib-specific effect\n",
    "- No significant hits for synergistic effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efc400e2-90b6-4b6c-8ada-a6c07f922bc2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Load dependencies\n",
    "\n",
    "Sys.setenv(LANGUAGE = \"en\") # set language to \"ja\" if you prefer\n",
    "suppressWarnings(library(DESeq2))\n",
    "#suppressWarnings(library(plyranges))\n",
    "suppressWarnings(library(dplyr))\n",
    "suppressWarnings(library(tibble))\n",
    "suppressWarnings(library(readr))\n",
    "suppressWarnings(library(ggplot2))\n",
    "suppressWarnings(library(ggrepel))\n",
    "suppressWarnings(library(extrafont))\n",
    "suppressWarnings(library(svglite))\n",
    "suppressWarnings(library(patchwork)) # combine plots\n",
    "\n",
    "suppressMessages(extrafont::font_import(pattern=\"Arial\",prompt=FALSE))\n",
    "suppressMessages(extrafont::loadfonts())\n",
    "\n",
    "getwd()\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0e015c-ff1c-4d90-9dc4-b297175466dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting defaults\n",
    "base_theme <- theme_classic(base_size=14, base_family=\"Arial\",) +\n",
    "    theme(axis.text = element_text(size=14,colour=\"black\"),\n",
    "          aspect.ratio=1,\n",
    "          #plot.margin=unit(c(0,0,0,0), \"null\")\n",
    "         )\n",
    "theme_set(base_theme)\n",
    "\n",
    "write_plot <- function(plt,outfile,width,height){\n",
    "    b=basename(outfile)\n",
    "    d=dirname(outfile)\n",
    "    dir.create(d, recursive=TRUE, showWarnings = FALSE)\n",
    "    pdf.options(encoding='ISOLatin2.enc')\n",
    "    #pdfName = paste(outfile, \".pdf\", sep=\"\")\n",
    "    pngName = paste(b, \".png\", sep=\"\")\n",
    "    svgName = paste(b, \".svg\", sep = \"\")\n",
    "    #ggsave(path=\"figures\", filename=pdfName, device=\"pdf\", width=width, height=height, units='in')\n",
    "    ggsave(path=d, device=\"png\", filename=pngName, width=width, height=height, units='in')\n",
    "    ggsave(path=d, device=\"svg\", filename=svgName, width=width, height=height, units='in')\n",
    "\n",
    "}\n",
    "\n",
    "# plotPCA without having to specify groups\n",
    "plot_pca <- function(dds,pcs=c(1,2),n=500,w=6,h=6){\n",
    "    vsd <- vst(dds, blind=TRUE)\n",
    "    rv <- rowVars(assay(vsd))\n",
    "    select <- order(rv, decreasing = TRUE)[seq_len(min(n, length(rv)))]\n",
    "    pca <- prcomp(t(assay(vsd)[select, ]))\n",
    "    percentVar <- pca$sdev^2 / sum(pca$sdev^2)\n",
    "    \n",
    "    df <- as.data.frame(pca$x)\n",
    "    df$name <- colnames(vsd)\n",
    "    \n",
    "    a <- ggplot(df, aes(.data[[paste0(\"PC\", pcs[1])]],\n",
    "                        .data[[paste0(\"PC\", pcs[2])]])) +\n",
    "      geom_point() +\n",
    "      geom_text_repel(aes(label = name), vjust = -0.5) +\n",
    "      xlab(paste0(\"PC\",pcs[1],\": \", round(percentVar[pcs[1]] * 100), \"% variance\")) +\n",
    "      ylab(paste0(\"PC\",pcs[2],\": \", round(percentVar[pcs[2]] * 100), \"% variance\")) #+\n",
    "      #coord_equal()\n",
    "    \n",
    "    #a = plotPCA(vsd, intgroup=c('DS1001b','vorasidenib','UNC0379')) + geom_label_repel(aes(label = name))\n",
    "    #write_plot(a,outfile='results/deseq2/shSHTN1-shL1CAM/pca',width=w,height=h)\n",
    "    options(repr.plot.width=w, repr.plot.height=h)\n",
    "    return(a)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9754f54c-1e71-4079-8e8c-a0001142cc1a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "load_gex  <- function(data_location){\n",
    "    message(\"loading gene expression data from \", data_location, \"...\")\n",
    "    cts = read.table(data_location,check.names=FALSE)\n",
    "    return(cts)\n",
    "}\n",
    "load_metadata <- function(data_location){\n",
    "    message(\"loading sample metadata from \", data_location, \"...\")\n",
    "    df = read.table(data_location,header=TRUE) %>%\n",
    "        mutate(IDH1i = factor(DS1001b | vorasidenib),\n",
    "               IDH2i = factor(vorasidenib),\n",
    "               KMT5Ai = factor(UNC0379),\n",
    "               treatment = case_when(\n",
    "                   (DS1001b | vorasidenib) & UNC0379 ~ 'combination',\n",
    "                   (DS1001b | vorasidenib) ~ 'IDHi',\n",
    "                   UNC0379 ~ 'KMT5Ai',\n",
    "                   TRUE ~ 'control'\n",
    "        ))\n",
    "    return(df)\n",
    "}\n",
    "\n",
    "counts_path = '../data/20260113/MGG152_hg19_countDATA_260106.txt' # 28263 genes\n",
    "other_counts_path = '../data/20260113/MGG152_hg19_countDATA_sel_260106.txt' # 12918 genes, subset of expressed genes running DESeq2 under default parameters.\n",
    "metadata_path = '../data/20260113/rna_sample_metadata.tsv'\n",
    "\n",
    "metadata <- load_metadata(metadata_path)\n",
    "counts <- load_gex(counts_path)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c641381d-f617-4978-96f6-6e09b31f8a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_deseq_data_set <- function(counts,metadata,design){\n",
    "    dds <- DESeq2::DESeqDataSetFromMatrix(\n",
    "        countData = counts,\n",
    "        colData = metadata,\n",
    "        design = design\n",
    "    )\n",
    "    smallestGroupSize <- 3\n",
    "    keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize\n",
    "    dds <- dds[keep,]\n",
    "    dds <- DESeq(dds)\n",
    "    return(dds)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19345462-2917-4f54-80f3-ca60cfb0ddaa",
   "metadata": {},
   "source": [
    "## Multifactor design\n",
    "Linear model using all samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d8a0e0-4bf8-4687-a041-f25106d7a200",
   "metadata": {},
   "outputs": [],
   "source": [
    "design = ~ IDH1i + IDH2i + KMT5Ai + IDH1i:KMT5Ai\n",
    "dds <- create_deseq_data_set(counts,metadata,design)\n",
    "dds\n",
    "resultsNames(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55c69d29-52fd-45ee-aff1-3fb828c3debd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_pca(dds)\n",
    "w=8;h=8\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "vsd <- vst(dds, blind=TRUE)\n",
    "a=plotPCA(vsd,intgroup='treatment',ntop=2000) + geom_label_repel(aes(label = name))\n",
    "write_plot(a,outfile='results/pca_all',width=w,height=h)\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "424902c2-615f-463b-9031-8678cceed9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "design = ~ IDH1i + IDH2i + KMT5Ai + IDH1i:KMT5Ai\n",
    "m <- metadata %>% filter(sample != 'MGG152_UNC1')\n",
    "dds <- create_deseq_data_set(\n",
    "    counts = counts %>% select(all_of(m$sample)),\n",
    "    metadata=m,design=design)\n",
    "dds\n",
    "resultsNames(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7477c56-35cf-44a3-9c6a-ee3b8daabf5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_pca(dds)\n",
    "w=8;h=8\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "vsd <- vst(dds, blind=TRUE)\n",
    "a=plotPCA(vsd,intgroup='treatment',ntop=2000) + geom_label_repel(aes(label = name))\n",
    "write_plot(a,outfile='results/pca_drop_outliers',width=w,height=h)\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e1c3f61-fe33-4c2e-a5fd-d5de77f8d9b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# After dropping outliers, PC1 is strongly associated with sequencing depth\n",
    "x=sizeFactors(dds)\n",
    "x[order(unlist(x))] %>% as_tibble\n",
    "names(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a0ded0-5a56-4fe0-80c5-0f2253e47d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = plot_pca(dds,pcs=c(2,3),n=2000)\n",
    "write_plot(a,outfile='results/pca_2_3',width=w,height=h)\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8a88f6-6d74-4ea4-8f4d-33324b360cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_syn = lfcShrink(dds,coef='IDH1iTRUE.KMT5AiTRUE',type='apeglm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "babebd89-8725-47ba-b59b-90894c219f3f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# (remove UNC1) Significant hits for synergistic effect: 0 at q < 0.1, 0 at p < 0.05\n",
    "syn_hits <- res_syn %>%\n",
    "    as.data.frame %>%\n",
    "    filter(padj < 0.1)\n",
    "syn_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14032879-9079-4dab-91e4-a195bd4a588a",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_idh = lfcShrink(dds,coef='IDH1i_TRUE_vs_FALSE',type='apeglm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62e62d9d-7392-439a-a684-1bed19d4cade",
   "metadata": {},
   "outputs": [],
   "source": [
    "# (remove UNC1) Significant hits for IDHi effect: 1 at q < 0.1, 1 at p < 0.05 (CBLN1)\n",
    "idh_hits <- res_idh %>%\n",
    "    as.data.frame %>%\n",
    "    filter(padj < 0.1)\n",
    "idh_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dacc0016-4ae3-41d8-9167-1aae23e50456",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_kmt5a = lfcShrink(dds,coef='KMT5Ai_TRUE_vs_FALSE',type='apeglm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2c772ae-8209-4a53-bfa5-10a70db96c3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# (remove UNC1) Significant hits for KMT5Ai effect: 1 at q < 0.1, 3 at p < 0.05 (H1-0)\n",
    "kmt5a_hits <- res_kmt5a %>%\n",
    "    as.data.frame %>%\n",
    "    filter(padj < 0.1)\n",
    "kmt5a_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03be0b8b-6147-4db5-801f-1a4fbc561313",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_vora = lfcShrink(dds,coef='IDH2i_TRUE_vs_FALSE',type='apeglm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c12a295f-0d05-4040-a0ac-822569858fc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# (remove UNC1) Significant hits for vorasidenib effect: 0 at q < 0.1, 0 at p < 0.05\n",
    "vora_hits <- res_vora %>%\n",
    "    as.data.frame %>%\n",
    "    filter(padj < 0.1)\n",
    "vora_hits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c2bd84-d967-4750-8307-bc73b37dfc9e",
   "metadata": {},
   "source": [
    "## Pairwise designs  \n",
    "For comparison, models using only the 3 replicates in each pairwise comparison.  \n",
    "(Dead code)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb27acc3-e618-4189-b8c5-f223432c8c90",
   "metadata": {},
   "source": [
    "### IDHi samples vs control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdbf3d6d-8dbb-41db-a6fc-2625237d504d",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_idhi = metadata %>%\n",
    "    filter(treatment %in% c('IDHi','control')) %>%\n",
    "    mutate(treatment = relevel(factor(treatment),ref='control'))\n",
    "meta_idhi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b0b9869-7382-4058-bff7-e9b402afc81a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds_idhi <- create_deseq_data_set(\n",
    "    counts = counts %>% select(all_of(meta_idhi$sample)),\n",
    "    metadata = meta_idhi,\n",
    "    design = ~ treatment\n",
    ")\n",
    "resultsNames(dds_idhi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53524d39-103f-4c0c-a4f3-0db1a6a80cfc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "res_idhi_sub = lfcShrink(dds_idhi,coef='treatment_IDHi_vs_control',type='apeglm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "796c00f2-9e89-497c-9483-f90a23b62da2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# No significant hits\n",
    "res_idhi_sub %>%\n",
    "    as.data.frame %>%\n",
    "    filter(padj < 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bac0184-f020-4d7f-8225-031d03b05906",
   "metadata": {},
   "source": [
    "### Intersection of vorasidenib | DS1001b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7376d431-78c4-4a0f-a744-8517c806b999",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_ds = metadata %>%\n",
    "    filter(treatment =='control' | (treatment == 'IDHi' & DS1001b)) %>%\n",
    "    mutate(treatment = relevel(factor(treatment),ref='control'))\n",
    "meta_ds\n",
    "\n",
    "dds_ds <- create_deseq_data_set(\n",
    "    counts = counts %>% select(all_of(meta_ds$sample)),\n",
    "    metadata = meta_ds,\n",
    "    design = ~ treatment\n",
    ")\n",
    "resultsNames(dds_ds)\n",
    "\n",
    "res_ds = lfcShrink(dds_ds,coef='treatment_IDHi_vs_control',type='apeglm')\n",
    "res_ds %>%\n",
    "    as.data.frame %>%\n",
    "    filter(padj < 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1206a7b-6eb2-4fa9-8422-7b28d1d299f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_vi = metadata %>%\n",
    "    filter(treatment =='control' | (treatment == 'IDHi' & vorasidenib)) %>%\n",
    "    mutate(treatment = relevel(factor(treatment),ref='control'))\n",
    "meta_vi\n",
    "\n",
    "dds_vi <- create_deseq_data_set(\n",
    "    counts = counts %>% select(all_of(meta_vi$sample)),\n",
    "    metadata = meta_vi,\n",
    "    design = ~ treatment\n",
    ")\n",
    "resultsNames(dds_vi)\n",
    "\n",
    "res_vi = lfcShrink(dds_vi,coef='treatment_IDHi_vs_control',type='apeglm')\n",
    "res_vi %>%\n",
    "    as.data.frame %>%\n",
    "    filter(padj < 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4db7e671-aa9a-4202-85f2-c4fb1502858d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "differential-expression",
   "language": "R",
   "name": "differential-expression"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
